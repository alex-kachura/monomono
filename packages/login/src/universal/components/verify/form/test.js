import React from 'react';
import VerifyForm from '.';
import { render, cleanup, fireEvent } from 'react-testing-library';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import { AppProvider, Formik } from '@oneaccount/react-foundations';
import { DefaultThemeProvider } from '@beans/theme';

expect.extend({ toMatchDiffSnapshot });

describe('VerifyForm component', () => {
  afterEach(cleanup);

  const mockFields = [
    {
      type: 'text',
      name: 'digit11',
      id: 'digit11',
      label: '11th',
    },
      {
      type: 'text',
      name: 'digit12',
      id: 'digit12',
      label: '12th',
    },
    {
      type: 'text',
      name: 'digit13',
      id: 'digit13',
      label: '13th',
    }
  ];

  const mockStateToken = 'mock-state-token';
  const mockCsrf = 'mock-csrf';

  const mockSchema = {
    type: 'object',
    required: [],
    properties: {
      digit11: {
        type: 'integer',
        maxLength: 1,
        errorMessage: 'Please enter the requested digit',
      },
      digit12: {
        type: 'integer',
        maxLength: 1,
        errorMessage: 'Please enter the requested digit',
      },
      digit13: {
        type: 'integer',
        maxLength: 1,
        errorMessage: 'Please enter the requested digit',
      },
      digit14: {
        type: 'integer',
        maxLength: 1,
        errorMessage: 'Please enter the requested digit',
      },
    },
    additionalProperties: false,
    errorMessage: {
      required: {
        digit11: 'Please enter the requested digit',
        digit12: 'Please enter the requested digit',
        digit13: 'Please enter the requested digit',
        digit14: 'Please enter the requested digit',
      },
    },
  };

  const mockInitialErrors = {
    digit11: 'server side error message',
    digit12: 'server side error message',
    digit13: 'server side error message',
  };

  function renderForm(errors = {}) {
    return render(
      <AppProvider
        appConfig={{
          getLocalePhrase: (key) => key,
        }}
      >
        <DefaultThemeProvider>
          <Formik
            initialValues={{
              digit11: '',
              digit12: '',
              digit13: '',
            }}
            initialErrors={errors}
            validationJSONSchema={mockSchema}
          >
            <VerifyForm
              fields={mockFields}
              stateToken={mockStateToken}
              csrf={mockCsrf}
              getLocalePhrase={(key) => key} // eslint-disable-line
            />
          </Formik>
        </DefaultThemeProvider>
      </AppProvider>
    );
  }

  const fireChange = (ele, val) => fireEvent.change(ele, { target: { value: val } });
  const fireKeyDown = (ele, key) => fireEvent.keyDown(ele, { key });

  let noErrorsFragment;

  describe('no initial errors', () => {
    it('should render correctly', () => {
      const { asFragment } = renderForm();

      noErrorsFragment = asFragment();

      expect(noErrorsFragment).toMatchSnapshot();
    });
  });

  describe('has initial errors', () => {
    test('should render error state of form', () => {
      const { asFragment } = renderForm(mockInitialErrors);
      const hasErrorsFragment = asFragment();

      expect(noErrorsFragment).toMatchDiffSnapshot(hasErrorsFragment);
    });
  });

  describe('form keyboard interaction', () => {
    let form;
    let input1;
    let input2;
    let input3;
    let button;

    beforeEach(() => {
      form = renderForm();
      input1 = form.container.querySelector('#digit11');
      input2 = form.container.querySelector('#digit12');
      input3 = form.container.querySelector('#digit13');
      button = form.container.querySelector('button[type=submit]');
    });

    describe('page loads', () => {
      test('should focus on first input', () => {
        expect(input1).toHaveFocus();
      });
    });

    describe('type number in first input', () => {
      test('should focus on 2nd input', () => {
        fireChange(input1, '1');

        expect(input2).toHaveFocus();
      });
    });

    describe('type number in second input', () => {
      test('should focus on 3rd input', () => {
        fireChange(input2, '2');

        expect(input3).toHaveFocus();
      });
    });

    describe('type number in third input', () => {
      test('should focus on submit button', () => {
        fireChange(input3, '3');

        expect(button).toHaveFocus();
      });
    });

    describe('backspace on populated third input', () => {
      test('should keep focus on 3rd input', () => {
        fireChange(input3, '3');
        input3.focus();
        fireKeyDown(input3, 'Backspace');

        expect(input3).toHaveFocus();
      });
    });

    describe('backspace on empty third input', () => {
      test('should focus on 2nd input', () => {
        input3.focus();
        fireKeyDown(input3, 'Backspace');

        expect(input2).toHaveFocus();
      });
    });

    describe('backspace on empty second input', () => {
      test('should focus on 1st input', () => {
        input2.focus();
        fireKeyDown(input2, 'Backspace');

        expect(input1).toHaveFocus();
      });
    });

    describe('backspace on empty first input', () => {
      test('should keep focus on 1st input', () => {
        input1.focus();
        fireKeyDown(input1, 'Backspace');

        expect(input1).toHaveFocus();
      });
    });

    describe('type letter in first input', () => {
      test('should keep focus on 1st input', () => {
        fireChange(input1, 'a');

        expect(input1).toHaveFocus();
      });

      test('should be no change to input', () => {
        fireChange(input1, 'a');

        expect(input1.value).toEqual('');
      });
    });
  });




});


