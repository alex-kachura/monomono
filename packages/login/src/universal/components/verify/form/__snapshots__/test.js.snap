// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyForm component has initial errors should render error state of form 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -12,11 +12,11 @@
        >
          <div
            class=\\"base-components__BaseElement-sc-150tvch-0 styled__LinkContainer-wt4ift-3 iYRfeQ\\"
          >
            <legend
-             class=\\"base-components__BaseElement-sc-150tvch-0 styled__Legend-wt4ift-2 isVbKl\\"
+             class=\\"base-components__BaseElement-sc-150tvch-0 styled__Legend-wt4ift-2 gsqUaM\\"
            >
              pages.verify.digits-prompt 
            </legend>
          </div>
          <div
@@ -25,17 +25,17 @@
            <div
              class=\\"sc-EHOje kdKQiq\\"
              name=\\"clubcard-form-group\\"
            >
              <label
-               class=\\"sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 gToxYy\\"
+               class=\\"sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 felILI\\"
                for=\\"digit11\\"
              >
                11th
              </label>
              <input
-               class=\\"base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 eyeGHp\\"
+               class=\\"base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 iRrnCT\\"
                id=\\"digit11\\"
                maxlength=\\"1\\"
                name=\\"digit11\\"
                required=\\"\\"
                type=\\"text\\"
@@ -45,17 +45,17 @@
            <div
              class=\\"sc-EHOje kdKQiq\\"
              name=\\"clubcard-form-group\\"
            >
              <label
-               class=\\"sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 gToxYy\\"
+               class=\\"sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 felILI\\"
                for=\\"digit12\\"
              >
                12th
              </label>
              <input
-               class=\\"base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 eyeGHp\\"
+               class=\\"base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 iRrnCT\\"
                id=\\"digit12\\"
                maxlength=\\"1\\"
                name=\\"digit12\\"
                required=\\"\\"
                type=\\"text\\"
@@ -65,26 +65,31 @@
            <div
              class=\\"sc-EHOje kdKQiq\\"
              name=\\"clubcard-form-group\\"
            >
              <label
-               class=\\"sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 gToxYy\\"
+               class=\\"sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 felILI\\"
                for=\\"digit13\\"
              >
                13th
              </label>
              <input
-               class=\\"base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 eyeGHp\\"
+               class=\\"base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 iRrnCT\\"
                id=\\"digit13\\"
                maxlength=\\"1\\"
                name=\\"digit13\\"
                required=\\"\\"
                type=\\"text\\"
                value=\\"\\"
              />
            </div>
          </div>
+         <p
+           class=\\"base-components__BaseElement-sc-150tvch-0 beans-form-group__error-message typography___styledElement-sc-186l0ce-0 hFMOIf\\"
+         >
+           server side error message
+         </p>
        </fieldset>
        <input
          name=\\"state\\"
          type=\\"hidden\\"
          value=\\"mock-state-token\\""
`;

exports[`VerifyForm component no initial errors should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-iwsKbI igoYVT"
  >
    <form
      autocomplete="off"
      method="POST"
      novalidate=""
    >
      <fieldset
        class="base-components__RootElement-sc-150tvch-1 styled__FormGroupContainer-wt4ift-1 dImxlC"
      >
        <div
          class="base-components__BaseElement-sc-150tvch-0 styled__LinkContainer-wt4ift-3 iYRfeQ"
        >
          <legend
            class="base-components__BaseElement-sc-150tvch-0 styled__Legend-wt4ift-2 isVbKl"
          >
            pages.verify.digits-prompt 
          </legend>
        </div>
        <div
          class="beans-form-group__children"
        >
          <div
            class="sc-EHOje kdKQiq"
            name="clubcard-form-group"
          >
            <label
              class="sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 gToxYy"
              for="digit11"
            >
              11th
            </label>
            <input
              class="base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 eyeGHp"
              id="digit11"
              maxlength="1"
              name="digit11"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-EHOje kdKQiq"
            name="clubcard-form-group"
          >
            <label
              class="sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 gToxYy"
              for="digit12"
            >
              12th
            </label>
            <input
              class="base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 eyeGHp"
              id="digit12"
              maxlength="1"
              name="digit12"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-EHOje kdKQiq"
            name="clubcard-form-group"
          >
            <label
              class="sc-ifAKCX ceiQhl styled__StyledLabel-sc-28yvb4-0 gToxYy"
              for="digit13"
            >
              13th
            </label>
            <input
              class="base-components__BaseInput-sc-150tvch-2 sc-bxivhb dznvxk styled-ep7354-0 eyeGHp"
              id="digit13"
              maxlength="1"
              name="digit13"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
      </fieldset>
      <input
        name="state"
        type="hidden"
        value="mock-state-token"
      />
      <input
        name="_csrf"
        type="hidden"
        value="mock-csrf"
      />
      <button
        aria-disabled="false"
        class="styled__BaseButton-sc-3mnirm-0 sc-dnqmqq cDuauq styled__PrimaryButton-sc-3mnirm-2 evotEg"
        role="button"
        type="submit"
      >
        <span
          class="base-components__BaseElement-sc-150tvch-0 styled__InnerContainer-sc-3mnirm-1 lnNmDn"
        >
          <span
            class="base-components__BaseElement-sc-150tvch-0 styled__BaseSpan-sc-3mnirm-4 beans-button__text styled__PrimarySpan-sc-3mnirm-5 hrUcvf"
          >
            pages.verify.buttons.submit
          </span>
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
